### ISSUE TITLE
Now taking requests for international translations...

### ISSUE THREAD

**Issue by justintocci** [workflowproducts/postage#59]
Postage is in regular use in twelve countries now. That's just in the last thirty days.

1) We need to figure out what languages need to be added. Obviously, community requests will go to the top of the list.

2) We need to plan how we want to do this. There are several choices:
    -install option (gets set in the config file)
    -no install option but still gets set in the config file
    -menu choice (usually a flag dropdown on top right)
    -use default from database
    -use default from OS

We no longer have control over the download and install process due to our dependance on Electron. I'm thinking that may mean that we can't allow users to choose their language on install. (Nunzio please look into this.)

I'm only opening this up for now. We'll need to get some answers before we proceed. Users that want to see translations should post requests here.

**Comment by nunziotocci** [comment]
I found some information on this subject:

https://fraserxu.me/2015/09/18/translation-workflow-in-Electron-application/
https://github.com/daxtero/electron-gettext

**Comment by nunziotocci** [comment]
Actually, I just noticed that that project is LGPLv2.1, is that ok for us?

**Comment by justintocci** [comment]
probably not ok. I don't know. It's funny. I've had a dozen people ask about licensing to see if they could use it. I tell everyone we've got a few MIT and BSD parts but the rest is all ours and they can do whatever they like. If you need us to sign something we'll sign it, etc... Then nothing happens.

**Comment by justintocci** [comment]
REMINDER: Users that want to see translations should post requests for specific languages here.

**Comment by justintocci** [comment]
I'm going to close this issue until the community requests a language.

**Comment by stevenwinfield** [comment]
...and I was just about to suggest en_GB! :-)

Seriously though, how would you implement it? A bunch of UTF-8 encoded files, where each line is of the form:
"text in en_US" <whitespace> "text in target language"

...or something similar?

If you give me all instances of "text in en_US" I don't mind writing the trivial en_GB translation, then anyone wanting to provide a translation in another language can easily fork your repo, then copy and modify that example, then issue a pull request.

**Comment by justintocci** [comment]
To tell you the truth, when I saw the map of users was international I panicked because the user base seemed evenly split between English speaking countries and others. I was mortified that they would think we were selfish for not having translations. 

I immediately posted the issue and expected users to request their native languages. It's a significant amount of work so we waited for the first request. 

I suspect that the reason it hasn't been requested is because the user interface is pretty minimal. I expect a lot of the words are likely jargon or loan words (and hence the same) in other languages. 

I'll reopen the issue for now, but I remain steadfast that we will not spend time on this until there is a request. We are a small company with very limited resources.

**Comment by justintocci** [comment]
@stevenwinfield, in reply to your question, I don't know. I looked into it a bit but I'm still open.

If I had to do this today then here is what we would do:

1) Create a new i18n component, maybe <gs-label> or <gs-i18n>.
2) Replace plain text used as labels in Postage with the new component. Set the name="label_text" attribute to the appropriate text for the label and context. The text inside the component would be the fallback text.
3) Add a preference to "Postage Options" for an international translation file. This would allow us to load the chosen version of the HTML document fragment into a javascript variable on load.
4) The i18n file format would be an HTML fragment. We would create an EN_US document as a reference. The format would consist of a CSS style tag and then a series of HTML elements. 

How it works:
All elements would have an ID attribute and could optionally reference classes declared in the accompanied CSS (to make it easy to do things like vertical text). The ID attributes would generally be underbar versions of the English reference translation but in the case where a label is used more than once, these would be broken out so that international users can display context aware translations.

At the top of a page, the i18n function would find the user's preference and load the i18n file into a javascript variable (or some faster thing). Once that is done, an event would be emitted, "i18n available". The i18n components would listen to this event and each find their appropriate replacement by searching with getElementById using the value inside their name= attribute.

If anyone has some advice we'd love to hear it.

**Comment by Vonatzki** [comment]
Hi, just stumbled upon this simple and wonderful application from your Reddit thread! Somewhat of a newbie postgresql user here.

Our database is using LATIN-1 encoding. I am not sure if this is caused by our database name being capitalized (name is "SCV").

Whenever I try to log in through the database it causes an error:

**_There has been an error with the Database connection.

..\..\db_framework_pq\db_framework.c:escape_conninfo_value: escape_conninfo_value: strlen = 0
Connect failed: FATAL:  database "scv" does not exist_**

Trying to get into 'template_01' and 'postgres' dummy databases works out just fine.

Thanks! So far I'm loving the interface!

**Comment by nunziotocci** [comment]
@Vonatzki It looks like we are lower-casing the DB name, so this is unrelated. Please open another issue so everything is separated.

**Comment by justintocci** [comment]
We're cleaning up today. I'm going to close this issue until the community requests a language.
